q2_demux/_demux.py

```diff
@@ -242,14 +244,19 @@ def _write_metadata_yaml(dir_fmt):
 
 def emp_single(seqs: BarcodeSequenceFastqIterator,
                barcodes: qiime2.CategoricalMetadataColumn,
+               golay_error_correction: bool = True,
```

Comment:
What if we go the route of `split_libraries_fastq`, and instead add in the equivalent of the `--max_barcode_errors` parameter? Then users would be able to disable error correct entirely, if we made this parameter something like:

```python
Int % Range(1, None) | Str % Choices(['disable'])
```

caveat --- that union isn't possible just yet, waiting on https://github.com/qiime2/qiime2/pull/444, which should go in this week. Either way though, seems like folks will want to be able to adjust the error threshold, right?

Reply:
I believe that is an open question whether, for example, two nucleotide errors in the barcode is indicative of more error in the read than if a single error is observed. Present approach would allow for an after the fact filtering (if desired) based on the stats output. If the there is a relationship between barcode errors and read errors, which may suggest exposing this filtering, then it may still be advantageous to retain the reads associated with error corrected barcodes so that downstream steps (e.g., DADA2) have a comprehensive perspective of error rather than a filtered one. 

I did overlook that parameter though :) good catch. The default in QIIME1 is `1.5`. All this being said, I'm not opposed to the modification :) just worried about what it would mean and whether it is still a good idea.

@nbokulich, do you know if it was the PHRED qual filtering work that informed `--max_barcode_errors` in QIIME1 by chance?

Reply:
> do you know if it was the PHRED qual filtering work that informed --max_barcode_errors in QIIME1 by chance?

The PHRED qual filtering work did not touch `max_barcode_errors` (as far as I recall). Thanks for checking!

Reply:
In light of some issues with the new types (re: parsing), I would avoid delaying anything on it. We can make a new PR when its ready. I don't have a good estimate on when that PR is going to be merged anymore :(

Reply:
I don't have strong feelings about which interface (the existing one, or the one suggested by @thermokarst) we go with. I think simplifying the interface for Golay correction by not giving the user control over `max_barcode_errors` but rather only over whether correction is applied or not is fine, and given @ebolyen's comment, we should probably just start there. I think it's possible that no one ever modified that parameter setting in QIIME 1. 

I do think it makes sense for the behavior to match that of QIIME 1 (i.e., barcode reads with more than 1.5 barcode errors are not corrected). Here's my reasoning. @wasade, I think you're right that no one knows if more barcode errors means a lower quality sequence read. I don't think that was the purpose of this parameter though, but I may be wrong about that (I wasn't involved in the development of this code - was it @justin212k who wrote this?). I thought that the idea was that the more barcode errors there are, the higher the chance of an incorrect "correction" (i.e., a read getting assigned to the wrong sample), which we definitely want to avoid as much as possible. If that's why this parameter was set the way it was, or if we don't know, I think we should match that QIIME 1 behavior unless we have a reason not to.

Reply:
Thanks, @gregcaporaso! I'm wondering right now if the default `max_barcode_errors` value in QIIME1 was possibly a carry over from the 454 days and whether it still is correct as this was the default in both `split_libraries_fastq.py` and `split_libraries.py`. From the Golay module doc, it states that "all <= 3 bit errors are corrected". An important point here is that the amount of error depends on the [type of substitution](https://github.com/biocore/qiime/blob/master/qiime/golay.py#L28-L29): "A <--> C" is 2 bit, "A <--> T" is 1 bit, for instance. For 8nt Hamming, correction looks like it was only applied if it was a [single bit error occurred](https://github.com/biocore/qiime/blob/master/qiime/hamming.py#L82), so the `max_barcode_errors` of 1.5 would have removed those sequences. 

So I guess I'm not sure if we do want to match the QIIME1 parameter value as it may be too conservative for 12nt Golay codes. 

I agree that we likely want filter out things > 3 bit errors.

Reply:
A little more literature mining, and found the following statement in [Hamady and Knight 2009 Genome Research](https://genome.cshlp.org/content/19/7/1141.full): "For example, using Hamming codes of eight bases can detect all double-bit errors and correct all single-bit errors (Hamady et al. 2008), and Golay codes of 12 bases can correct all triple-bit errors and detect all quadruple-bit errors." 

On this, I do think we want and drop out sequences with > 3 bit errors. I'll set it to continue to record those dropped in the statistics file. Commit coming in a moment.

Reply:
This sounds great @wasade, thanks for the detective work on this! It'll be great to finally have Golay correction in QIIME 2!